server/util.go

```diff
@@ -543,3 +549,71 @@ func (h CorsHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 	h.handler.ServeHTTP(w, req)
 }
+
+// newLogWrapped creates a RecordSet that logs the result of the underlying RecordSet.
+func newLogWrapped(rs sqlexec.RecordSet, maxCachedRows int, fields ...zap.Field) sqlexec.RecordSet {
+	newRs := &recordSetLoggerWrapper{
+		rs:            rs,
+		maxCachedRows: maxCachedRows,
+		cachedRows:    make([]chunk.Row, 0),
+		zapFields:     fields,
+	}
+	return newRs
+}
+
+type recordSetLoggerWrapper struct {
+	rs            sqlexec.RecordSet
+	cachedRows    []chunk.Row
+	zapFields     []zap.Field
+	maxCachedRows int
+}
+
+func (r *recordSetLoggerWrapper) Fields() []*ast.ResultField {
+	return r.rs.Fields()
+}
+
+func (r *recordSetLoggerWrapper) Next(ctx context.Context, req *chunk.Chunk) error {
+	chk := r.rs.NewChunk(nil)
```

Comment:
It looks to me that the better way is to directly read to `req` and then get some rows from it, but it seems kind of difficult to know when `req` is reset? 🤔

Reply:
Yeah that's not something we can control.